### DISCUSSION TITLE
[2023/05/23] Frontend Night

### DISCUSSION THREAD

**Discussion by jjenko** [wantedly/frontend-night#258] [General]
### 2023/05/23のテーマ: Google I/O 2023 で発表されたもの

- 資料の準備をがんばらない
- 参加・途中退席は自由
- スキップしない

話したい・聞きたいネタを書いてく ✏️

ハッシュタグ [#frontend_night](https://twitter.com/hashtag/frontend_night?src=hash)

Prev: https://github.com/wantedly/frontend-night/discussions/257 

 _Created from https://github.com/wantedly/frontend-night/discussions/179 by [issue-creator](https://github.com/rerost/issue-creator)_

**Comment by chloe463** [comment]
- [What's new in Web](https://youtu.be/x9rh0Du4Czg)
- [What's new in web UI](https://youtu.be/buChHSdsF9A)
- [Getting started with Angular Signals](https://youtu.be/EEzDLpIbW9w)

**Comment by NotFounds** [comment]
## [Our top Core Web Vitals recommendations for 2023](https://web.dev/top-cwv-2023/)

- LCP(Largest Contentful Paint) が大事
  - 特に画像
    - 画像自体を高速に配信、読み込むことが大事だが画像の読み込み開始までの時間を短縮することも大事
    - `img` タグの `src` や `srcset` 属性を使う
      - `data-src` はダメ
    - CSR より SSR を優先
  - `fetchpriority="high"` という属性を指定するとリソースの読み込み優先順位をブラウザに教えられる
    - 現在(2023/01)Chromiumベースのブラウザのみでサポートされているが他のブラウザでは無視されるのでつけておくのが良い
  - LCP リソースには `loading="lazy"` をつけない
- CLS(Cumulative Layout Shift)
  - 画像の layout shift を防ぐには `width` と `height` を指定しておく
  - CSS の `aspect-ratio` も重要
    - 画面サイズに基づいてブラウザが高さを計算できる
  - `min-height` を指定することでデフォルトで指定される `0px` よりは影響を軽減できる
  - `bfcache`(back/forward cache) の対象であるかを確認することが重要
    - devtools からテストできるっぽい
  - `top` や `left` を使って絶対指定でアニメーションするのではなく `transform:translateX()` や `transform:translateY()` を使うことでアニメーション時の layout shift を抑えられる
    - GPU 上で動作するのでパフォーマンス改善にもつながる
- FID(First Input Delay)
  - FID には欠点があり改善の余地がある
  - FID は今後 Interaction to Next Paint(INP) に置き換わるかも
  - 50ms を超えるような長いタスクはなるべく行わないことでメインスレッドを空けることができる
  - `isInputPending` や `Schedular API` を検討する
  - 不要な JavaScript を削減する
  - 大きな DOM 変更を避ける

**Comment by NotFounds** [comment]
## [What's new in Web](https://www.youtube.com/watch?v=x9rh0Du4Czg)

### HTML の `modal`
```html
<dialog id='d'>
  <form method='dialog'>
    <p>Hi, I'm a dialog.</p>
    <button>ok</button>
  </form>
</dialog>

<button onclick='d.showModal()'>
  Open Dialog.
</button>
```
フォーカス管理やタブ移動、スタッキングコンテキストなどをブラウザに任せられる

### CSS transform
プロパティごとに個別に記述できるようになったので `@keyframes` と組み合わせることで簡単に複雑なアニメーションが実装できる

### ViewPort
モバイル端末のアドレスバーを考慮することができる
`sv(h|w|min|max)` , `lv(h|w|min|max)`

### structuredClone
goodbye👋 `JSON.parse(JSON.stringfy(original))`
あまりに使われるので V8 は最適化まで行っていた

### focus-visible
フォーカスを受け取ったときの疑似要素

### TransformStream
すべてのブラウザでサポートされた
`readable` と `writable` のペアになったやつ

### importmap
```html
<script type='importmap'>
  {
    "imports": "https://unpkg.com/lodash@4.17.21/lodash.js",
    "util": "./modules/util.js"
  }
</script>
<script type='module'>
  import _ from 'lodash';
  console.log(_.map([1, 2, 3], (n) => n * 2*));
</script>
```
外部モジュール名のマッピングを対応する URL に定義できる

### baseline
https://web.dev/baseline/
ブラウザがサポートしている機能をまとめていく

**Comment by NotFounds** [comment]
## [What's new in web UI](https://www.youtube.com/watch?v=buChHSdsF9A)

### Container queries
- すべてのブラウザで安定
- メディアクエリに比べてより細かい単位で指定できる
  - viewport 基準ではなく任意の要素のサイズが基準になる

### Style queries
- 親要素のスタイルに応じて子要素のスタイルを変えられる
- `@container style(--rain >= 75%)` みたいなことができる

### `:has` selector
- 親の要素にある子要素が含まれているかを判断できる

### `:nth-child` of type
- `:has(:nth-child(n + 11))` みたいなことができる
- `:nth-child(2 of .special)`

### `text-wrap:` balance
- バランスよく見出し小見出しを改行してくれる

### initial-letter
- 頭文字のドロップキャップ 

### Viewport units
- モバイル端末のアドレスバーを考慮した高さ

### Wide-gamut color
- 色空間が広がった
  - もとは sRGB

### `color-mix()`
- 2つの色を混ぜることができる
- `color-mix(in var(--color-space), var(--color-1), var(--color-2))`
- 透明度も混ぜられる

### CSS Nesting
- SaaS にはあったがデフォルトでサポートされたっぽい？

### Cascade layer

### Scoped styles
- CSS Native ネームスペースを作れる

### Trig Funcitons
- 三角関数

### Individual transforms
- 個別の%ごとに transform を指定できる

### Popover
- ポップアップ UI がネイティブで使えるようになった

### Anchor positioning
- tooltip とかが作りやすくなりそう

### Selectmenu
- select 要素ではできなかった表現ができる？
- button とか入れられるの嬉しい

### Discrete animations

### Scroll-driven animations
- スクロールとアニメーションを連携するやつ

### View transitions

**Comment by NotFounds** [comment]
後で見たい
## [Debugging modern web applications](https://www.youtube.com/watch?v=3lNkq264nkM)
